client/components/data/query-sites/index.jsx

```diff
@@ -17,66 +13,74 @@ import { requestSites, requestSite } from 'state/sites/actions';
 import { getPreference } from 'state/preferences/selectors';
 import getPrimarySiteId from 'state/selectors/get-primary-site-id';
 
-class QuerySites extends Component {
-	UNSAFE_componentWillMount() {
-		this.requestAll( this.props );
-		this.requestPrimary( this.props );
-		this.requestRecent( this.props );
-		this.requestSingle( this.props );
-	}
-
-	UNSAFE_componentWillReceiveProps( nextProps ) {
-		if ( nextProps.siteId !== this.props.siteId ) {
-			this.requestSingle( nextProps );
-		}
+const getRecentSites = state => getPreference( state, 'recentSites' );
 
-		if ( nextProps.primaryAndRecent && nextProps.primarySiteId !== this.props.primarySiteId ) {
-			this.requestPrimary( nextProps );
-		}
+const requestAll = () => ( dispatch, getState ) => {
+	if ( ! isRequestingSites( getState() ) ) {
+		dispatch( requestSites() );
+	}
+};
 
-		if (
-			nextProps.primaryAndRecent &&
-			! isEqual( nextProps.recentSiteIds, this.props.recentSiteIds )
-		) {
-			this.requestRecent( nextProps );
-		}
+const requestSingle = siteId => ( dispatch, getState ) => {
+	if ( siteId && ! isRequestingSite( getState(), siteId ) ) {
+		dispatch( requestSite( siteId ) );
 	}
+};
 
-	requestAll( props ) {
-		if ( props.allSites && ! props.requestingSites ) {
-			props.requestSites();
-		}
+const requestPrimary = siteId => ( dispatch, getState ) => {
+	if ( siteId && ! hasAllSitesList( getState() ) && ! isRequestingSite( getState(), siteId ) ) {
```

Comment:
Can you call `getState()` just once here and especially in `requestRecent`, where it's called inside a `some` loop?

Reply:
Sure thing! Didn't realise that would be a de-opt.